### PR TITLE
[DPE-4739] Avoid using time.sleep in rollback integration tests

### PR DESCRIPTION
**Pull request by shayancanonical**
## Issue
1. We use time.sleep instead of a deterministic way to determine when to send `resume-upgrade` after rolling back. This sometimes leads to [this error](https://github.com/canonical/mysql-router-k8s-operator/actions/runs/10312040824/job/28546894902#step:31:2522) when `resume-upgrade` is run too early
```
Resume upgrade event failed: Highest number unit is unhealthy. Upgrade will not resume.
```
2. We are using `prime` in charmcraft which is deprecated in charmcraft 3.1.1

## Solution
1. Use a more deterministic approach - check that all units are `active` and `idle` (one will have rolled back while the other two will not)
2. Use parts to prime version files and the dump plugin to prime scripts/ directory

### CODE REVIEW

File: charmcraft.yaml [hunk 1]
```diff
@@ -10,6 +10,14 @@ bases:
...
     architectures: [arm64]
 parts:
+  version_files:
+    plugin: nil
+    prime:
+      - charm_version
+      - workload_version
+  scripts:
+    plugin: dump
+    source: scripts/
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply charmcraft.yaml:20]: i agree on the need for a unified approach. however, it's very annoying to have to sync my charmcraft version with the version of CI and to have a local build process that's different than the build process in CI

My thinking: use `dump` for directories, use `nil` with `prime` for individual files

File: tests/integration/test_upgrade.py [hunk 2]
```diff
@@ -197,7 +196,10 @@ async def test_fail_and_rollback(ops_test: OpsTest, continuous_writes) -> None:
     logger.info("Re-refresh the charm")
     await mysql_router_application.refresh(path="./upgrade.charm")
 
-    time.sleep(30)
+    await ops_test.model.block_until(
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tests/integration/test_upgrade.py]: good idea, addressed in [d531e71](https://github.com/canonical/mysql-router-k8s-operator/pull/298/commits/d531e710acc1c1518781940ee555a2bb198050e3)

File: charmcraft.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (shayancanonical) [reply charmcraft.yaml]: ok, i think we're converging on a strategy. i think the way forward is always have one part for `files`. if we need to copy a directory, we use `plugin: dump` with `source: directory/`. if no directory needs to be imported, use `plugin: nil`

individual files are `prime`d in the `files` part

**Review by taurus-forever — APPROVED** [review]
We are still pending on sleep() duration guess. But as you wish.

File: tests/integration/test_upgrade.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulomach) [reply tests/integration/test_upgrade.py]: Did you tested without any sleep and an extra `block_until` instead to detect something like `lambda: not all(unit.workload_status == "active" for unit in mysql_router_application.units)`

**Review by paulomach — COMMENTED** [review]
Left a question only

File: tests/integration/test_upgrade.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tests/integration/test_upgrade.py]: the issue is that the charm may be in `active` (if the refresh is not immediate) or `maintenance` (if the refresh starts before the test code executes). there's no way to know, and so i think a sleep is justified here - we just ensure refresh starts and then check for the correct conditions to ensure we can run `resume-upgrade`

File: charmcraft.yaml [hunk 3]
```diff
@@ -10,6 +10,12 @@ bases:
     channel: "22.04"
     architectures: [arm64]
 parts:
+  files:
+    plugin: dump
+    source: scripts/
+    prime:
+      - charm_version
+      - workload_version
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply charmcraft.yaml]: addressed in c9b9b4f19a6a53427b71e53e6e0e13f92ce19a96

**Pull request merged**